Monika After Story/game/script-affection.rpy

```diff
@@ -249,6 +299,699 @@ init -900 python in mas_affection:
+        """
+        global __is_dirty
+        __is_dirty = True
+        __set_pers_data(get_default_data())
+
+    def __get_data():
+        """
+        Returns current data (decoded),
+        ALWAYS use this accessor
+
+        OUT:
+            - tuple with the data
+            - None if an error happened
+        """
+        data = __get_pers_data()
+        if data is None:
+            mas_utils.mas_log.critical("Aff data is invalid")
+            return None
+
+        data = __decode_data(data)
+        if data is None:
+            mas_utils.mas_log.critical("Failed to decode aff data")
+            return None
+
+        return data
+
+    def _get_aff():
+        """
+        Private getter that handles errors,
+        you should probably use public version
+
+        OUT:
+            float - current affection
+        """
+        data = __get_data()
+        if data is None:
+            return 0.0
+
+        return data[0]
+
+    def _get_today_cap():
+        """
+        Returns today's aff cap
+
+        OUT:
+            tuple[float, float] - current aff cap
+        """
+        data = __get_data()
+        if data is None:
+            return (0.0, 0.0)
+
+        return data[2:4]
+
+    def __validate_timestamp(ts, now_ts):
+        """
+        Verifies the given time against current time
+
+        IN:
+            ts - the timestamp to validate
+            now_ts - the current time
+
+        OUT:
+            float:
+                original timestamp if it's valid
+                or modified timestamp
+        """
+        # If you didn't time travel, I wouldn't need to fix this
+        delta_t = ts - now_ts
+        hour_t = 3600
+        day_t = hour_t * 24
+        timezone_hop = hour_t * 30
+
+        if delta_t > timezone_hop:# 30h for timezone changes
+            log.info("INVALID TIME, FREEZING TIME")
+            penalty = max(min(delta_t, day_t*30), day_t)
+            ts = now_ts + penalty
+
+        return ts
+
+    def _grant_aff(amount, bypass, reason=None):
+        """
+        Grants some affection
+
+        IN:
+            amount - float - amount of affection to grant
+            bypass - bool - is this bypass gain or not
+            reason - str/None - the reason for this bonus,
+                MUST be current topic label or None
+                (Default: None)
+        """
+        global __is_dirty
+
+        data = __get_data()
+        if not data:
+            return
+        data = list(data)
+
+        now_ = time.time()
+        data[4] = __validate_timestamp(data[4], now_)
+
+        freeze_date = datetime.date.fromtimestamp(data[4])
+        if store.mas_pastOneDay(freeze_date):
+            data[2] = 0.0
+            data[3] = 0.0
+            data[4] = now_
+
+        frozen = data[2] >= 7.0
+
+        # Sanity checks
+        amount = float(amount)
+        if amount <= 0.0:
+            raise ValueError("Invalid value for affection: {}".format(amount))
```

Comment:
Generally, don't mind if we just log the error, but I want you to consider this again because if we screw up, we would know it right away - we would get an exception while testing. If it'd just get logged, we may miss it entirely.